server/src/main/java/org/candlepin/resteasy/AnnotationLocator.java

```diff
@@ -0,0 +1,141 @@
+        }
+
+        List<Binding<?>> rootResourceBindings = new ArrayList<>();
+        for (final Binding<?> binding : injector.getBindings().values()) {
+            final Type type = binding.getKey().getTypeLiteral().getType();
+            if (type instanceof Class) {
+                final Class<?> beanClass = (Class) type;
+                if (GetRestful.isRootResource(beanClass)) {
+                    rootResourceBindings.add(binding);
+                }
+            }
+        }
+        for (Binding<?> binding : rootResourceBindings) {
+            Class<?> clazz = (Class) binding.getKey().getTypeLiteral().getType();
+            if (Proxy.isProxyClass(clazz)) {
+                for (Class<?> intf : clazz.getInterfaces()) {
+                    ResourceClass resourceClass = ResourceBuilder.rootResourceFromAnnotations(intf);
+                    registerConcreteMethods(resourceClass, clazz);
+                }
+            }
+            else {
+                ResourceClass resourceClass = ResourceBuilder.rootResourceFromAnnotations(clazz);
+                registerConcreteMethods(resourceClass, clazz);
+            }
+        }
+        lock();
+    }
+
+    private void registerConcreteMethods(ResourceClass resourceClass, Class<?> concreteClass) {
+        for (ResourceMethod resourceMethod : resourceClass.getResourceMethods()) {
+            registerConcreteMethod(resourceMethod.getMethod(), concreteClass);
+        }
+        for (ResourceLocator resourceMethod : resourceClass.getResourceLocators()) {
+            registerConcreteMethod(resourceMethod.getMethod(), concreteClass);
+        }
+    }
+
+    private void registerConcreteMethod(Method method, Class<?> concreteClass) {
+        try {
+            for (Class<?> iface : concreteClass.getInterfaces()){
+                Method ifaceMethod = iface.getMethod(method.getName(), method.getParameterTypes());
+                internalMap.put(ifaceMethod, method);
+            }
+        }
+        catch (NoSuchMethodException e) {
+            log.error("Can't find concrete method for method {}!", method);
+        }
+    }
+
+    /**
+     * Looks up annotation for a resource from the concrete implementation.
+     *
+     * Falls back to any instance of that annotation on the method itself if no concrete implementation is
+     * registered.
+     *
+     * @param method method to get annotation for
+     * @param annotationClass which annotation to try to get
+     */
+    public <T extends Annotation> T getAnnotation(Method method, Class<T> annotationClass) {
+        Method concreteMethod = internalMap.get(method);
+        if (concreteMethod != null) {
+            T annotation = concreteMethod.getAnnotation(annotationClass);
+            if (annotation != null) {
+                return annotation;
+            }
+        }
+        return method.getAnnotation(annotationClass);
+    }
+
+    public synchronized void lock() {
```

Comment:
I imagine the ```synchronized``` keyword is going to move up to init and then this will become private. Is there any situation where we want this to be locked externally?

Reply:
Nope, I'll push an update shortly.